### PR TITLE
[FIX] mail: prevent load messages if already loading

### PR DESCRIPTION
**Pull request by xdu-odoo**
task-2373952

### CODE REVIEW

**Comment by xdu-odoo** [comment]
(Original PR targeted master  https://github.com/odoo/odoo/pull/61166)

File: addons/mail/static/src/models/thread_cache/thread_cache.js [hunk 1]
```diff
@@ -267,6 +267,9 @@ function factory(dependencies) {
          * @returns {mail.message[]}
          */
         async _loadMessages({ extraDomain, limit = 30 } = {}) {
+            if (this.isLoading) {
+                return;
+            }
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread_cache/thread_cache.js:272]: This does not feel good, it's already prevented in `loadMoreMessages` and `loadNewMessages`. And the compute of `hasToLoadMessages` should be taking care of setting itself to `true` only when it's appropriate to actually load. If there is a fix, it should be in that compute I think, or in one of the fields the compute is using, maybe `isCacheRefreshRequested`.

**Comment by xdu-odoo** [comment]
I think I found a logic error. If not, then I don't understand the condition in its original form.

**Comment by xdu-odoo** [comment]
@seb-odoo 
Is it possible to test the search view with QUnit or do I have to write a tour ?

**Comment by seb-odoo** [comment]
I think the new control panel has some qunit helpers, but I never looked into them.

**Comment by xdu-odoo** [comment]
Where could I find these ?

**Comment by xdu-odoo** [comment]
Nevermind, found them

File: addons/mail/static/src/models/thread_cache/thread_cache.js [hunk 2]
```diff
@@ -203,8 +203,8 @@ function factory(dependencies) {
             if (!wasCacheRefreshRequested && this.threadViews.length === 0) {
                 // don't load message that won't be used
                 return false;
-                }
-            if (!wasCacheRefreshRequested && (this.isLoaded || this.isLoading)) {
+            }
+            if ((!wasCacheRefreshRequested && this.isLoaded) || this.isLoading) {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread_cache/thread_cache.js:207]: Shouldn't it just be `||` between all parts of this condition?

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread_cache/thread_cache.js:207]: IMO, if the thread cache is loaded then we need to do the message fetch if a cache refresh is requested.
Otherwise if this is already loading then we don't need to do the same request again.
That's how I understand this

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_cache/thread_cache.js:207]: Ok what you just said sounds correct indeed, so the only difference is if we have a request but it is already loading, we ignore the request. I wonder if it can lead to issues or not, but I can't think of any so I guess that's fine.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_cache/thread_cache.js:207]: Split the condition then, preventive return if loading on its own line.

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread_cache/thread_cache.js:207]: As it's the same thread cache you don't have any domain change so normally the current request is exactly the same so it seems natural to cancel the new one. To be sure we could cancel the old one and make the new but I think this will affect performances and not change anything at the end

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 3]
```diff
@@ -4262,6 +4268,31 @@ QUnit.test('mark channel as seen if last message is visible when switching chann
...
+            if (args.method === 'message_fetch') {
+                const domainsAsStr = args.kwargs.domain.map(domain => domain.join(''));
+                assert.step(`message_fetch_${domainsAsStr.join('')}`);
+            }
+            return this._super(...arguments);
+        },
+    });
+    assert.verifySteps(['message_fetch_needaction=true'], "A message_fetch request should have been done for needaction messages");
+
+    // Open filter menu of control panel and select a custom filter (id = 0)
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4289]: what is id=0? where does it come from

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 4]
```diff
@@ -4262,6 +4268,31 @@ QUnit.test('mark channel as seen if last message is visible when switching chann
...
+    await toggleFilterMenu(document.body);
+    await toggleAddCustomFilter(document.body);
+    await applyFilter(document.body);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4292]: not your fault but those are very unclear and the api is strange too (`document.body`?!)

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Overall this is ok but the test is hard to follow (too much magic going on, I don't understand where values are coming from)

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 5]
```diff
@@ -4262,6 +4268,31 @@ QUnit.test('mark channel as seen if last message is visible when switching chann
...
+    assert.verifySteps(['message_fetch_id=0needaction=true'], "A message_fetch request should have been done for selected filter");
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4293]: also where does `needaction=true` come from? worth a comment that it is coming from inbox being selected by default, I assume?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4289]: It's the only custom filter available

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 4 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4292]: It needs an element to work, so I gave body ..

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 5 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4293]: You resolved this and pushed changes but I didn't see answer to my message or any explanation added in code?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4292]: Ok, I think we are meant to send `this.widget` just in case there are somehow multiple control panels on the page, but I guess `body` is fine too since at the end we always use `body` in our new tests so that's consistent with what we do.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 5 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4293]: I changed the assert comment : 

```js
assert.verifySteps(['message_fetch_needaction=true'], "A message_fetch request should have been done for needaction messages as inbox is selected by default");
```

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 6]
```diff
@@ -4262,6 +4268,31 @@ QUnit.test('mark channel as seen if last message is visible when switching chann
...
+
+    this.data['mail.channel'].records.push({
+        id: 20,
+        name: "General"
+    });
+    await this.start({
+        async mockRPC(route, args) {
+            if (args.method === 'message_fetch') {
+                const domainsAsStr = args.kwargs.domain.map(domain => domain.join(''));
+                assert.step(`message_fetch_${domainsAsStr.join('')}`);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4282]: would split `message_fetch` from the rest with another separator (`:`?) for clarity of assert, and maybe joint the rest with `,`?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 5 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4293]: Ok I see it's posted as initial assert actually, I would still clarify comment here that it should be selected filter + domain of current thread (or however we can call that)

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Any chance we could we add a custom filter in test and use that one instead?

Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Ideally should test result of filter (= matching message in message list) instead of mock rpc.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
